### ISSUE TITLE
Reissue: “[New Title]” — inherits from pyright#10750 (incorrect generic attribute assignment)

### ISSUE THREAD

**Issue by akoshel** [microsoft/pyright#10777]
This issue reissues [microsoft/pyright#10750](https://github.com/microsoft/pyright/issues/10750) to track the same scenario in a new context.
The original discussion concluded that the current behavior is by design, but I’d like to document the case again for visibility and potential reconsideration in future updates.

Summary of Scenario
When assigning to a generic attribute constrained by a TypeVar, valid assignments that match the constraints are still rejected.
```python
from typing import Generic, TypeVar
from abc import ABC, abstractmethod


T1 = TypeVar('T1')

class Base(ABC, Generic[T1]):
    a: T1

    @abstractmethod
    def some_method(self) -> T1:
        ...

class Derived1(Base[int]):

    def __init__(self, p: bool) -> None:
        self.a = 1

    def some_method(self) -> int:
        return self.a

    

T2 = TypeVar('T2', int, str)

class Derived2(Base[T2]):

    def __init__(self, p: bool) -> None:
        if p:
            self.a = 1
        else:
            self.a = 'string'

    def some_method(self) -> T2:
        return self.a

```

```
30:22 - error: Cannot assign to attribute "a" for class "Derived2[T2@Derived2]*"
    Type "Literal[1]" is not assignable to type "T2@Derived2" (reportAttributeAccessIssue)


32:22 - error: Cannot assign to attribute "a" for class "Derived2[T2@Derived2]*"
    Type "Literal['string']" is not assignable to type "T2@Derived2" (reportAttributeAccessIssue)
2 errors, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Pyright's behavior here is correct, so I don't consider this a bug. Not surprisingly, mypy reports the same type violations for this code.

The code in `Derived2` is unsound. Consider if you called `Derived2[str](True)`.

I recommend against using a value-constrained type variable for situations like this.
